crates/solver/src/solver/score_computation.rs

```diff
@@ -51,7 +51,16 @@ impl ScoreCalculator {
             - self.gas_price_factor * gas_price / 10_000_000_000.
             - self.nmb_orders_factor * nmb_orders as f64;
         let success_probability = 1. / (1. + exponent.exp());
-        let score = success_probability * (surplus + fees - gas_amount * gas_price);
+        let obj = surplus + fees - gas_amount * gas_price;
+        let score: f64;
+        let cap = 10000000000000000.0;
+        if success_probability * obj <= cap && (1.0 - success_probability) * obj <= cap {
+            score = success_probability * obj;
+        } else if success_probability * obj > cap && success_probability >= 0.5 {
+            score = obj - cap * (1.0 - success_probability) / success_probability;
+        } else {
+            score = cap * success_probability / (1.0 - success_probability);
+        }
```

Comment:
A slightly more "Rust-y" way to write this is:
```suggestion
        let cap = 10000000000000000.0;
        let score = if success_probability * obj <= cap && (1.0 - success_probability) * obj <= cap {
            success_probability * obj
        } else if success_probability * obj > cap && success_probability >= 0.5 {
            obj - cap * (1.0 - success_probability) / success_probability
        } else {
            cap * success_probability / (1.0 - success_probability)
        };
```